File: src/gwaspeek/plot_state.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Dict

import numpy as np
import pandas as pd


DEFAULT_BUILD = "37"
CANONICAL_CHR_LENGTHS: dict[str, dict[int, float]] = {
    "37": {
        1: 249_250_621.0,
        2: 243_199_373.0,
        3: 198_022_430.0,
        4: 191_154_276.0,
        5: 180_915_260.0,
        6: 171_115_067.0,
        7: 159_138_663.0,
        8: 146_364_022.0,
        9: 141_213_431.0,
        10: 135_534_747.0,
        11: 135_006_516.0,
        12: 133_851_895.0,
        13: 115_169_878.0,
        14: 107_349_540.0,
        15: 102_531_392.0,
        16: 90_354_753.0,
        17: 81_195_210.0,
        18: 78_077_248.0,
        19: 59_128_983.0,
        20: 63_025_520.0,
        21: 48_129_895.0,
        22: 51_304_566.0,
        23: 155_270_560.0,
        24: 59_373_566.0,
        25: 16_569.0,
    },
    "38": {
        1: 248_956_422.0,
        2: 242_193_529.0,
        3: 198_295_559.0,
        4: 190_214_555.0,
        5: 181_538_259.0,
        6: 170_805_979.0,
        7: 159_345_973.0,
        8: 145_138_636.0,
        9: 138_394_717.0,
        10: 133_797_422.0,
        11: 135_086_622.0,
        12: 133_275_309.0,
        13: 114_364_328.0,
        14: 107_043_718.0,
        15: 101_991_189.0,
        16: 90_338_345.0,
        17: 83_257_441.0,
        18: 80_373_285.0,
        19: 58_617_616.0,
        20: 64_444_167.0,
        21: 46_709_983.0,
        22: 50_818_468.0,
        23: 156_040_895.0,
        24: 57_227_415.0,
        25: 16_569.0,
    },
}


@dataclass(frozen=True)
class GenomeLayout:
    build: str
    offsets: Dict[int, float]
    chr_sizes: Dict[int, float]
    x_min: float
    x_max: float


@dataclass(frozen=True)
class PlotDataset:
    df: pd.DataFrame
    layout: GenomeLayout
    x: np.ndarray
    chrom: np.ndarray
    pos: np.ndarray
    p: np.ndarray
    mlog10p: np.ndarray


def _coerce_chr_max_pos(raw_map: object) -> Dict[int, float] | None:
    if not isinstance(raw_map, dict):
        return None
    out: Dict[int, float] = {}
    for chrom, max_pos in raw_map.items():
        try:
            out[int(chrom)] = float(max_pos)
        except (TypeError, ValueError):
            continue
    return dict(sorted(out.items())) if out else None


def normalize_build(build: str | None) -> str:
    token = str(build or DEFAULT_BUILD).strip().upper()
    aliases = {
        "37": "37",
        "38": "38",
        "GRCH37": "37",
        "GRCH38": "38",
        "HG19": "37",
        "HG38": "38",
    }
    return aliases.get(token, DEFAULT_BUILD)


def build_genome_layout(
    df: pd.DataFrame,
    build: str | None = None,
    *,
    data_driven_lengths: bool = False,
) -> GenomeLayout:
    resolved_build = normalize_build(build or str(df.attrs.get("genome_build") or DEFAULT_BUILD))
    chr_max_pos = _coerce_chr_max_pos(df.attrs.get("chr_max_pos"))
    if chr_max_pos is None:
        chr_max = df.groupby("CHR")["POS"].max().sort_index()
        chr_max_pos = {int(chrom): float(max_pos) for chrom, max_pos in chr_max.items()}
    if not chr_max_pos:
        raise ValueError("No variants available to build genome layout.")

    canonical = CANONICAL_CHR_LENGTHS.get(resolved_build, {})
    offsets: Dict[int, float] = {}
    chr_sizes: Dict[int, float] = {}
    offset = 0.0
    for chrom, max_pos in chr_max_pos.items():
        offsets[chrom] = offset
        if data_driven_lengths:
            span = max(1.0, float(max_pos))
        else:
            span = float(canonical.get(chrom, max_pos))
        chr_sizes[chrom] = span
        offset += span

    sorted_chroms = sorted(chr_sizes)
    first_chrom = sorted_chroms[0]
    last_chrom = sorted_chroms[-1]
    x_min = float(offsets[first_chrom])
    x_max = float(offsets[last_chrom] + chr_sizes[last_chrom])

    if len(df) == 0:
        return GenomeLayout(build=resolved_build, offsets=offsets, chr_sizes=chr_sizes, x_min=x_min, x_max=x_max)

    return GenomeLayout(
        build=resolved_build,
        offsets=offsets,
        chr_sizes=chr_sizes,
        x_min=x_min,
        x_max=x_max,
    )


def prepare_plot_dataset(
    df: pd.DataFrame,
    layout: GenomeLayout | None = None,
    build: str | None = None,
    *,
    data_driven_lengths: bool = False,
) -> PlotDataset:
    resolved_layout = layout or build_genome_layout(
        df, build=build, data_driven_lengths=data_driven_lengths
    )
    if len(df) == 0:
        x = np.empty(0, dtype=float)
        chrom = np.empty(0, dtype=int)
        pos = np.empty(0, dtype=float)
        p = np.empty(0, dtype=float)
        mlog10p = np.empty(0, dtype=float)
    else:
        x = df["POS"].to_numpy(dtype=float, copy=False) + df["CHR"].map(resolved_layout.offsets).to_numpy(dtype=float)
        chrom = df["CHR"].to_numpy(dtype=int, copy=False)
        pos = df["POS"].to_numpy(dtype=float, copy=False)
        p = df["P"].to_numpy(dtype=float, copy=False)
        mlog10p = df["mlog10p"].to_numpy(dtype=float, copy=False)
    return PlotDataset(
        df=df,
        layout=resolved_layout,
        x=x,
        chrom=chrom,
        pos=pos,
        p=p,
        mlog10p=mlog10p,
    )


def visible_mask(dataset: PlotDataset, x_start: float, x_end: float) -> np.ndarray:
    if len(dataset.x) == 0:
        return np.zeros(0, dtype=bool)
    return (dataset.x >= float(x_start)) & (dataset.x <= float(x_end))

File: rust/gwaspeek/src/plot_state.rs
use std::collections::{BTreeMap, HashMap};
use std::sync::OnceLock;

use crate::preprocess::CleanSumstats;

pub const DEFAULT_BUILD: &str = "37";

fn canonical_chr_lengths(build: &str) -> &'static HashMap<i32, f64> {
    static M37: OnceLock<HashMap<i32, f64>> = OnceLock::new();
    static M38: OnceLock<HashMap<i32, f64>> = OnceLock::new();
    match build {
        "38" => M38.get_or_init(|| {
            let mut m = HashMap::new();
            m.insert(1, 248_956_422.0);
            m.insert(2, 242_193_529.0);
            m.insert(3, 198_295_559.0);
            m.insert(4, 190_214_555.0);
            m.insert(5, 181_538_259.0);
            m.insert(6, 170_805_979.0);
            m.insert(7, 159_345_973.0);
            m.insert(8, 145_138_636.0);
            m.insert(9, 138_394_717.0);
            m.insert(10, 133_797_422.0);
            m.insert(11, 135_086_622.0);
            m.insert(12, 133_275_309.0);
            m.insert(13, 114_364_328.0);
            m.insert(14, 107_043_718.0);
            m.insert(15, 101_991_189.0);
            m.insert(16, 90_338_345.0);
            m.insert(17, 83_257_441.0);
            m.insert(18, 80_373_285.0);
            m.insert(19, 58_617_616.0);
            m.insert(20, 64_444_167.0);
            m.insert(21, 46_709_983.0);
            m.insert(22, 50_818_468.0);
            m.insert(23, 156_040_895.0);
            m.insert(24, 57_227_415.0);
            m.insert(25, 16_569.0);
            m
        }),
        _ => M37.get_or_init(|| {
            let mut m = HashMap::new();
            m.insert(1, 249_250_621.0);
            m.insert(2, 243_199_373.0);
            m.insert(3, 198_022_430.0);
            m.insert(4, 191_154_276.0);
            m.insert(5, 180_915_260.0);
            m.insert(6, 171_115_067.0);
            m.insert(7, 159_138_663.0);
            m.insert(8, 146_364_022.0);
            m.insert(9, 141_213_431.0);
            m.insert(10, 135_534_747.0);
            m.insert(11, 135_006_516.0);
            m.insert(12, 133_851_895.0);
            m.insert(13, 115_169_878.0);
            m.insert(14, 107_349_540.0);
            m.insert(15, 102_531_392.0);
            m.insert(16, 90_354_753.0);
            m.insert(17, 81_195_210.0);
            m.insert(18, 78_077_248.0);
            m.insert(19, 59_128_983.0);
            m.insert(20, 63_025_520.0);
            m.insert(21, 48_129_895.0);
            m.insert(22, 51_304_566.0);
            m.insert(23, 155_270_560.0);
            m.insert(24, 59_373_566.0);
            m.insert(25, 16_569.0);
            m
        }),
    }
}

/// Exposed for tests mirroring `tests/test_preprocess.py`.
pub fn canonical_length(build: &str, chrom: i32) -> Option<f64> {
    canonical_chr_lengths(build).get(&chrom).copied()
}

#[derive(Debug, Clone)]
pub struct GenomeLayout {
    pub build: String,
    pub offsets: BTreeMap<i32, f64>,
    pub chr_sizes: BTreeMap<i32, f64>,
    pub x_min: f64,
    pub x_max: f64,
}

#[derive(Debug, Clone)]
pub struct PlotDataset {
    pub layout: GenomeLayout,
    pub x: Vec<f64>,
    pub chrom: Vec<i32>,
    pub pos: Vec<f64>,
    pub p: Vec<f64>,
    pub mlog10p: Vec<f64>,
}

pub fn normalize_build(build: Option<&str>) -> String {
    let token = build.unwrap_or(DEFAULT_BUILD).trim().to_uppercase();
    match token.as_str() {
        "37" | "GRCH37" | "HG19" => "37".to_string(),
        "38" | "GRCH38" | "HG38" => "38".to_string(),
        _ => DEFAULT_BUILD.to_string(),
    }
}

pub fn build_genome_layout(
    clean: &CleanSumstats,
    build: Option<&str>,
    data_driven_lengths: bool,
) -> Result<GenomeLayout, String> {
    let resolved_build = normalize_build(build);
    let chr_max_pos = &clean.chr_max_pos;
    if chr_max_pos.is_empty() {
        return Err("No variants available to build genome layout.".to_string());
    }

    let canonical = canonical_chr_lengths(resolved_build.as_str());

    let mut offsets: BTreeMap<i32, f64> = BTreeMap::new();
    let mut chr_sizes: BTreeMap<i32, f64> = BTreeMap::new();
    let mut offset = 0.0_f64;
    for (&chrom, &max_pos) in chr_max_pos.iter() {
        offsets.insert(chrom, offset);
        let span = if data_driven_lengths {
            max_pos.max(1.0)
        } else {
            *canonical.get(&chrom).unwrap_or(&max_pos)
        };
        chr_sizes.insert(chrom, span);
        offset += span;
    }

    let sorted_chroms: Vec<i32> = chr_sizes.keys().copied().collect();
    let first_chrom = *sorted_chroms.first().unwrap();
    let last_chrom = *sorted_chroms.last().unwrap();
    let x_min = *offsets.get(&first_chrom).unwrap();
    let x_max = offsets.get(&last_chrom).unwrap() + chr_sizes.get(&last_chrom).unwrap();

    Ok(GenomeLayout {
        build: resolved_build,
        offsets,
        chr_sizes,
        x_min,
        x_max,
    })
}

pub fn prepare_plot_dataset(
    clean: &CleanSumstats,
    build: Option<&str>,
    data_driven_lengths: bool,
) -> Result<PlotDataset, String> {
    let layout = build_genome_layout(clean, build, data_driven_lengths)?;
    let n = clean.chr.len();
    if n == 0 {
        return Ok(PlotDataset {
            layout,
            x: vec![],
            chrom: vec![],
            pos: vec![],
            p: vec![],
            mlog10p: vec![],
        });
    }
    let mut x = Vec::with_capacity(n);
    for i in 0..n {
        let c = clean.chr[i];
        let off = *layout.offsets.get(&c).unwrap();
        x.push(clean.pos[i] + off);
    }
    Ok(PlotDataset {
        layout,
        x,
        chrom: clean.chr.clone(),
        pos: clean.pos.clone(),
        p: clean.p.clone(),
        mlog10p: clean.mlog10p.clone(),
    })
}

pub fn visible_mask(dataset: &PlotDataset, x_start: f64, x_end: f64) -> Vec<bool> {
    if dataset.x.is_empty() {
        return vec![];
    }
    dataset
        .x
        .iter()
        .map(|xv| *xv >= x_start && *xv <= x_end)
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::io::load_sumstats;
    use crate::preprocess::preprocess_sumstats;
    use std::path::PathBuf;

    fn fixture(name: &str) -> PathBuf {
        PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .join("../../tests/fixtures")
            .join(name)
    }

    #[test]
    fn data_driven_uses_observed_max() {
        let raw = crate::preprocess::RawSumstats {
            chrom_tokens: vec!["23".into(), "23".into()],
            pos_tokens: vec!["5000000".into(), "8000000".into()],
            p_tokens: vec!["1e-8".into(), "1e-8".into()],
            mlog10p_tokens: vec![],
            has_p_column: true,
        };
        let clean = preprocess_sumstats(raw, 5.0).unwrap();
        let ds = prepare_plot_dataset(&clean, Some("37"), true).unwrap();
        assert_eq!(ds.layout.chr_sizes.get(&23).copied().unwrap(), 8_000_000.0);
        assert!(ds.layout.chr_sizes[&23] < canonical_length("37", 23).unwrap());
    }

    #[test]
    fn canonical_lengths_match_python_table() {
        let raw = load_sumstats(
            &fixture("sumstats_small.tsv"),
            "\t",
            None,
            None,
            None,
            None,
        )
        .unwrap();
        let clean = preprocess_sumstats(raw, 3.0).unwrap();
        let ds = prepare_plot_dataset(&clean, Some("37"), false).unwrap();
        assert_eq!(
            ds.layout.chr_sizes.get(&1).copied().unwrap(),
            canonical_length("37", 1).unwrap()
        );
    }
}
